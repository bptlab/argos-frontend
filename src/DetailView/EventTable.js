import React from "react";
import {connect, PromiseState} from "react-refetch";
import ConnectionComponent from "./../Utils/ConnectionComponent.js";
import {Table, TableBody, TableHeader, TableHeaderColumn, TableRow, TableRowColumn} from "material-ui/Table";
import EventTabs from "./EventTabs";
import FilterBar from "./../Utils/FilterBar";
import {css} from "aphrodite";
import AppStyles from "./../AppStyles";
import config from "./../config/config";

class EventTable extends ConnectionComponent {

	constructor(props) {
		super(props);

		this.state = {
			filter: [],
		};

		this.handleActiveEventTypeChange = this.handleActiveEventTypeChange.bind(this);
		this.handleFilterChange = this.handleFilterChange.bind(this);
	}

	composeTableHeader(eventTypeAttributes) {
		return (
			<TableHeader
				displaySelectAll={false}
				adjustForCheckbox={false}>
				<TableRow>
					{eventTypeAttributes.map(
						(attribute, key) => {
							return (
								<TableHeaderColumn
									className={css(AppStyles.capitalize)}
									key={key}>
									{attribute.Name}
								</TableHeaderColumn>
							);
						})
					}
				</TableRow>
			</TableHeader>
		);
	}

	composeTableRow(event, key) {
		const row = event.Attributes.map(
			(eventAttribute, attributeKey) => {
				return (
					<TableRowColumn key={attributeKey}>
						{eventAttribute.Value}
					</TableRowColumn>
				);
			});
		return (<TableRow key={key}>{row}</TableRow>);
	}

	composeTableBody(events) {
		return (
			<TableBody displayRowCheckbox={false}>
				{events.map((event, key) => {
					if (!this.isCoveredByFilter(event)) {
						return "";
					}
					return this.composeTableRow(event, key);
				})}
			</TableBody>
		);
	}

	isCoveredByFilter(event) {
		// every is equivalent to logical and over an array
		return this.state.filter.every((filter) => {
			return this.testFilter(event, filter);
		});
	}

	testFilter(event, filter) {
		if (!filter.value) {
			return true;
		}
		let columnsToBeSearched = this.props.eventTypeAttributes.value;
		if(filter.column) {
			columnsToBeSearched = columnsToBeSearched.filter((eventPropertyKey) => {
				return EventTable.doesContain(eventPropertyKey.Name, filter.column);
			});
		}

		return columnsToBeSearched.some((eventPropertyKey) => {
			const splittedFilterValues = filter.value.split(",");
			const eventAttribute = event.Attributes.find(attribute => attribute.Name === eventPropertyKey.Name);
			return splittedFilterValues.some(filterValue => {
				const currentFilterValue = filterValue.trim();
				if (currentFilterValue &&  EventTable.doesContain(eventAttribute.Value, currentFilterValue)) {
					return true;
				}
			});
		});
	}

	static doesContain(baseValue, subValue) {
		return (baseValue.toString().toLowerCase().indexOf(subValue.toString().toLowerCase()) > -1);
	}

	handleActiveEventTypeChange(eventTypeId) {
		this.setState({filter: []});
		this.props.lazyEventLoading(eventTypeId);
	}

	handleFilterChange(filter) {
		this.setState({
			filter: filter,
		});
	}

	loadTableContent() {
		const eventFetchingIncomplete = super.render(PromiseState.all([
			this.props.eventTypeAttributes,
			this.props.events]));
		if(eventFetchingIncomplete) {
			return eventFetchingIncomplete;
		}
		return (
			<Table>
				{this.composeTableHeader(this.props.eventTypeAttributes.value)}
				{this.composeTableBody(this.props.events.value)}
			</Table>
		);
	}

	loadFilterBar() {
		return (
			<FilterBar
				styles={[AppStyles.elementMarginTop]}
				onFiltersChange={this.handleFilterChange}
				autoCompleteSource={this.props.eventTypeAttributes.value.map(attributeInfo => attributeInfo.Name)} />
		);
	}

	render() {
		const eventTypes = this.props.eventTypes.value;
		const connectionIncomplete = super.render(this.props.eventTypes);
		if(connectionIncomplete) {
			return connectionIncomplete;
		}
		let tableContent = "";
		let filterBar = "";
		if (this.props.eventTypeAttributes && this.props.events) {
			tableContent = this.loadTableContent();
			if (this.props.eventTypeAttributes.value) {
				filterBar = this.loadFilterBar();
			}
		}
		return (
			<div>
				{filterBar}
				<EventTabs
					eventTypes={eventTypes}
					eventTypeChangeHandler={this.handleActiveEventTypeChange}
					styles={[AppStyles.elementMarginTop]} />
				{tableContent}
			</div>
		);
	}
}

export default connect.defaults({fetch: ConnectionComponent.switchFetch})(props => ({
	eventTypes: config.backendRESTRoute + `/entity/${props.entityId}/eventtypes`,
	lazyEventLoading: eventTypeId => ({
		eventTypeAttributes: config.backendRESTRoute + `/eventtype/${eventTypeId}/attributes`,
		events: config.backendRESTRoute + `/entity/${props.entityId}/eventtype/${eventTypeId}/events/0/10000`
	}),
}))(EventTable);
