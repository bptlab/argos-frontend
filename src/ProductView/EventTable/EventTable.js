import React, { Component } from 'react';
import EventTableHeader from './EventTableHeader/EventTableHeader.js';
import EventTableRow from './EventTableRow/EventTableRow.js';

class EventTable extends Component {
    
    testFilter(currentEvent) {
        // every is equivalent to logical and over an array
        return this.props.filter.every((filter) => {
            return this.testFilterArguments(currentEvent, filter);
        });
    }

    testFilterArguments(currentEvent, currentFilter) {
        if (!currentFilter.value) {
            return true;
        }
        let columnsToBeSearched = this.props.header;
        if(currentFilter.column) {
          columnsToBeSearched = columnsToBeSearched.filter((eventPropertyKey) => {
            return this.doesContain(eventPropertyKey.name, currentFilter.column);
          });
        }

        return columnsToBeSearched.some((eventPropertyKey) => {
            const splittedFilterValues = currentFilter.value.split(",");
            for (var i = 0; i < splittedFilterValues.length; i++) {
                if (this.doesContain(currentEvent[eventPropertyKey.name], splittedFilterValues[i].trim())) {
                    return true;
                }
            }
        });
    }

    doesContain(baseValue, subValue) {
      return (baseValue.toString().toLowerCase().indexOf(subValue.toString().toLowerCase()) > -1);
    }

    render() {
        return (
            <div className="event-table-container">
                <table className="table">
                    <EventTableHeader eventTypeAttributes={this.props.header}/>
                    <tbody>
                    {this.props.events.map((event, index) => {
                        if(this.testFilter(event)) {
                            return (
                                <EventTableRow 
                                    key={index} 
                                    id={index} 
                                    event={event}
                                    eventTypeAttributes={this.props.header}/>
                            );
                        } else {
                            return false;
                        }
                    })}
                    </tbody>
                </table>
            </div>
        );
    }
} export default EventTable;
