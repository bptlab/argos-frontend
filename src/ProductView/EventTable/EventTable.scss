@import "EventTableHeader/EventTableHeader";
@import "EventTableRow/EventTableRow";

.event-table-container {
  overflow-x: auto;
}