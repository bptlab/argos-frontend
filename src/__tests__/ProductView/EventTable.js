import React from 'react';
import EventTable from '../../ProductView/EventTable/EventTable.js';
import renderer from 'react-test-renderer';
import TestEvents from '../testData/frontend_events.js';
import TestEventTypes from '../testData/frontend_eventTypes';
let instance;

test("Rendering of EventTable", () => {
    const mockFilter = {id: 'mock-filter', value: '4', column: 'errorcode'};
    const component = renderer.create(
        <EventTable ref={(child) => {instance = child}}
                    header={TestEventTypes.EVENTTYPES[0].attributes}
                    events={TestEvents.EVENTS}
                    filter={[mockFilter]}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Inclusion of two values", () => {
  expect(instance.doesContain("baseString", "string")).toBeTruthy();
  expect(instance.doesContain([123], "12")).toBeTruthy();
  expect(instance.doesContain("baseString", "strr")).toBeFalsy();
});
