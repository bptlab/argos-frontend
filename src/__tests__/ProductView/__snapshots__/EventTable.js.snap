exports[`test Rendering of EventTable 1`] = `
<div
  className="container event-table-container">
  <table
    className="table">
    <thead
      className="event-table-header thead">
      <tr>
        <th>
          #
        </th>
        <th>
          ErrorCode
        </th>
        <th>
          ErrorDescription
        </th>
        <th>
          PossibleFix
        </th>
        <th>
          DeviceID
        </th>
        <th>
          DeviceLocation
        </th>
        <th>
          SoftwareVersion
        </th>
        <th>
          DateOfServiceIntervention
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row">
          5
        </th>
        <td>
          404
        </td>
        <td>
          Product not found.
        </td>
        <td>
          Search for it.
        </td>
        <td>
          666
        </td>
        <td>
          Not known.
        </td>
        <td>
          0.0.4.alpha
        </td>
        <td>
          2016-11-02T00:00:00.000+0200
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
